Replace argh with argparse

stratakis opened the pull request:
This is an initial attempt to replace argh with the stdlib argparse library to resolve #398 

Should be compatible across all python's supported by barman (at least locally it was). The tests for the cli pass as well. However there are still some issues to resolve, like the help command showing properly the positional arguments.

Let me know what you think of that approach.

stratakis commented:
Using aliases for python2 is non-trivial but it should be possible by utilizing something like https://gist.github.com/sampsyo/471779 which is based on the initial patchset for [CPython](https://bugs.python.org/issue9234)  adding aliases to argparse. I'll see if I can integrate it with the current approach.

stratakis commented:
> Thank you @stratakis for this PR - the overall approach looks reasonable and is definitely heading in the right direction.
> 
> There are a couple of blocking issues:
> 
>     * The barman commands which rely on `barman.__config__` being available are currently failing, e.g.:
> 
> 
> ```
> barman show-servers all
> EXCEPTION: 'NoneType' object has no attribute 'server_names'
> See log file for more details.
> 'NoneType' object has no attribute 'server_names'
> See log file for more details.
> Traceback (most recent call last):
>   File "/Users/michael.wallace/src/EnterpriseDB/barman/barman/cli.py", line 1495, in main
>     args.func(args)
>   File "/Users/michael.wallace/src/EnterpriseDB/barman/barman/cli.py", line 669, in show_servers
>     servers = get_server_list(args)
>   File "/Users/michael.wallace/src/EnterpriseDB/barman/barman/cli.py", line 1350, in get_server_list
>     available_servers = barman.__config__.server_names()
> AttributeError: 'NoneType' object has no attribute 'server_names'
> ```
> 

This should hopefully be fine after @jthreefoot-edb's recommendation but please verify if it looks ok.

>     * We would lose the singular forms of `show-servers`, `list-servers` and `show-backups` in python 2 - it's unfortunate argparse doesn't support the `aliases` keyword argument for python 2 but we need to keep those singular forms for backward compatibility.
> 

Working on it per https://github.com/EnterpriseDB/barman/pull/435#issuecomment-956858179

> 
> Also a couple of smaller issues:
> 
>     * There are still a few references to argh in the comments of the `pretty_args` down on [L1250](https://github.com/EnterpriseDB/barman/pull/435/files#diff-d91ba4117ac85c960aa4252ff9b8df565954676cf8148c1eb07cde33f2734ee4L1250).
> 

Done.

>     * argh added a `help` subcommand implicitly which does the same as what running `barman` on its own does with these changes - it would be nice if we could still have `barman help` just to keep the CLI consistent with previous versions.
>

I'll check that out as well.
 
> 
> If we can get those two blockers resolved then we should be able to get this merged and say goodbye to argh.

stratakis commented:
Aliases should now work for python2.

mikewallace1979 commented on barman/cli.py:
```diff
@@ -52,6 +54,116 @@
 
+# Support aliases for argparse in python2
+if sys.version_info.major < 3:
+    class AliasedSubParsersAction(_SubParsersAction):
```
Nice hack and thank you for putting the python2 workaround in.

This looks like it was mostly derived from https://gist.github.com/sampsyo/471779 - if so it's worth referencing that gist in a comment and including a note that the code is public domain. Although the CC0 1.0 license doesn't require attribution I think it's still important to note where the code came from just in case there are some issues with automated code scans in the future.

mikewallace1979 commented:
Another thing I just noticed testing locally is that the `diagnose` command is gone - it didn't have any argh decorators but it's the `cli.diagnose` function. It doesn't take any arguments so probably just needs a `@command()` decorator.

stratakis commented:
A help command was added. Since I've found no easy way unfortunately to add an alias to --help I've simply created a new subparser which prints out the help. However this does not work the same as argh:

`$ barman help` will print out the help but only if we are under the barman user, since the global_config is called

argh gave the possibility to also do something like `$ barman help list-servers` which would be equivalent to `$ barman list-servers --help`. That is not the case with the new subparser adding the `help` command. Possibly I'll have to delve again into the private attributes or argparse, however the complexity seems quite big if the argh compatibility is to be maintained 100%.

mikewallace1979 commented:
> Possibly I'll have to delve again into the private attributes or argparse, however the complexity seems quite big if the argh compatibility is to be maintained 100%.

Agreed - this might be somewhere we need to break compatibility for the sake of keeping complexity under control. Given that the `help` command is nearly always going to be used interactively (I can't think of any use cases for running help in scripts) it doesn't seem too bad to change behaviour here - and the `--help` option is always there to provide subcommand-specific help when needed.

mikewallace1979 commented on barman/cli.py:
```diff
@@ -67,8 +186,11 @@ def check_target_action(value):
+@command([
+        argument("--minimal",
+                 help="machine readable output")
+        ],
```
I think we need an `action="store_true"` here (there wasn't one in the original code, I suppose argh must have been adding that automatically).

mikewallace1979 commented on barman/cli.py:
```diff
@@ -67,8 +186,11 @@ def check_target_action(value):
+                 help="machine readable output")
+        ],
+        cmd_aliases=["list-server"])
 def list_servers(minimal=False):
```
This should accept `args` and then down on L208 we can reference `args.minimal`. Currently its getting the args namespace object as the `minimal` argument which is truthy and therefore forcing the minimal output even if the option isn't set.

stratakis replied:
Hope I understood it correctly, check the latest commit.

mikewallace1979 replied:
Yep, looks good :+1:

mikewallace1979 commented on barman/cli.py:
```diff
@@ -691,42 +731,28 @@ def show_servers(args):
+                  metavar="TIMEOUT",
+                  default="30",
+                  type=check_non_negative)],
+         cmd_aliases=["switch_xlog"])
```
This should be `switch-xlog`

mikewallace1979 reviewed (commented):
A few extra comments following testing locally - this is looking pretty good though! We'll give it a final review tomorrow.

stratakis commented:
Some more caveats that I've noticed when comparing the new and old help output side by side:

Running `--help` now will show the dict of the positional arguments including the aliases and in a non-alphabetical order. argh would sort the arguments and would not show the aliases.

Also argh would show a list of all the subparsers as positional arguments on the help text along with their help. In the case of argparse this is for explicitly set arguments on the main parser, but not for the subparsers.

And I haven't looked yet on the shell completion with arghcomplete if things work there.

mikewallace1979 commented on barman/cli.py:
```diff
@@ -790,6 +804,7 @@ def check(args):
 
 
+@command()
 def diagnose():
```
Just noticed this still needs a placeholder arg as argparse is going to give it a namespace.

stratakis commented:
By the way I'll be squashing the commits to one when everything is reviewed and ready.

didiermichel commented on barman/cli.py:
```diff
@@ -52,6 +54,127 @@
+    """
+    def decorator(func):
+        parser = parent.add_parser(func.__name__.replace('_','-'), description=func.__doc__,
+                                   help=func.__doc__, aliases=cmd_aliases)
```
I spent time on  add the help section yesterday but I realized you just add it :) Nice job.

didiermichel reviewed (commented):
Hi @stratakis, 
I had a look at the code. I like the decorator.

You should try run `tox -e black` to format your code before merging.

didiermichel commented on barman/cli.py:
```diff
@@ -790,7 +808,8 @@ def check(args):
 
-def diagnose():
+@command()
+def diagnose(args):
```
Also, maybe rebase on master to resolve PR conflict (`tests/requirements_minimal.txt`).
This will allow CI to run tests.

jthreefoot-edb replied:
> Hi, does we need args as it is not used(or maybe in decorator)?
> It also this breaks tests/test_diagnose.py where diagnose() is used.
> maybe set args=[]

It would need to be `args=None` actually. [https://docs.python-guide.org/writing/gotchas/](https://docs.python-guide.org/writing/gotchas/) 

I don't think this particular case would ever cause an actual problem bc diagnose doesn't have any arguments, but may as well set it to None.

mikewallace1979 commented:
Something like this would fix ordering of the subcommands in both the positional arguments and the expanded help, though it is yet more coupling to argparse internals: https://gist.github.com/mikewallace1979/8d9ae2fa84ce5f4fc0b70999fbd5b04c

stratakis commented:
> Something like this would fix ordering of the subcommands in both the positional arguments and the expanded help, though it is yet more coupling to argparse internals: https://gist.github.com/mikewallace1979/8d9ae2fa84ce5f4fc0b70999fbd5b04c

This is quite elegant, the solution I was trying was far more complex than that. Added it.

stratakis commented:
Also another thing I noticed is that the doctext shown for each command when running `$ barman --help` is not properly wrapped on the terminal as it happens with argh, or argparse by default for that matter. A newline is added after ~50 chars instead of extending the line up to where the terminal is.

mikewallace1979 commented:
Giving this PR a final check now - we can get autocompletion working by adding `argcomplete.autocomplete(p)` just before the call to `p.parse_args` on L1786.

stratakis commented:
I'll squash the commits after the final review.

mikewallace1979 commented on barman/cli.py:
```diff
@@ -98,16 +254,13 @@ def list_servers(minimal=False):
+        )
+    ]
 )
 def cron(keep_descriptors=False):
```
This is another place we were relying on argh magic which I missed in the last review - we need to do the same change we did for `list-servers`, specifically: accept `args` instead of `keep_descriptors`, add `action="store_true"` the `--keep-descriptors` argument, and refer to `args.keep_descriptors` inside the `cron` function.

mikewallace1979 commented:
@stratakis Thanks for making all the changes. I'm going to push a copy of this branch directly to the EnterpriseDB repo so that the full CI pipeline can run (there are some behind-the-scenes integration which will take around 20 minutes) and if that all succeeds then I think we're good to squash and merge.

mikewallace1979 commented on barman/cli.py:
```diff
@@ -52,6 +56,156 @@
+    return (list(name_or_flags), completer, kwargs)
+
+
+def command(args=[], parent=subparsers, cmd_aliases=[]):
```
We should avoid using mutables as defaults for `args` and `cmd_aliases` here, as we did with `args=None` earlier in the PR.

mikewallace1979 replied:
Yes - we'd also need to test for `None` and assign an empty list for both parameters at the top of the function.

mikewallace1979 reviewed (commented):
One last round of feedback - sorry!

mikewallace1979 commented:
@stratakis Ok I think you're all good to squash this down to one commit ready for merging.

stratakis commented:
Squashed and rebased on top of master. Edited the commit message as well. Should I also add anything else, such as a changelog entry?

mikewallace1979 commented:
> Squashed and rebased on top of master. Edited the commit message as well. Should I also add anything else, such as a changelog entry?

I think we've got everything we need for this PR. We'll update the NEWS at release time and we retired the ChangeLog since it was just duplicating the commit history.

Thanks again for taking this on and sticking with it (and indeed identifying the issue in the first place) - it's great that we've been able to wave goodbye to unmaintained argh so quickly! 🎉🎉🎉

stratakis commented:
Thanks as well for reviewing the PR, addressing all the issues with it and merging it in such a swift manner :)

Is there an ETA for the next release?

mikewallace1979 commented:
The next release should be out by the end of this coming week - 19th November or earlier.
